### PR TITLE
Problem: No way to run ParsecT

### PR DESCRIPTION
**Pull request by cognivore**
Solution:

 - Unpacking of a built-up ParsecT, by design, is performed via unParser, into a M (Reply S E A),
    - which in turn holds consumption status (T|F),
    - final state (which we can obtain from within ParserT) and the result of parsing, which keeps a rolling result of type A.
 - Thus, we define four private functions, one for each case (Consumed/Empty | OK/Error), for other functions to be able to run after unpacking.

Sadly, that's not enough to define a low-level parser, thus...

Immediate further work:
 - Next step to complete prerequisites for writing low-level parsers is, seemingly, just defining MonadParsec class!

### CODE REVIEW

File: Megaparsec/Megaparsec.lean [hunk 1]
```diff
@@ -58,21 +60,37 @@ inductive Result (S E A: Type) [Stream S] where
...
+    -- Report errors with State into M B
     (ParseError S E -> State S E -> M B) ->              -- Consumed-Error
+    -- Return A with State S E and Hints into M B
     (A -> State S E -> Hints (Stream.Token S) -> M B) -> -- Empty-OK
+    -- Report errors with State into M B
     (ParseError S E -> State S E -> M B) ->              -- Empty-Error
     M B
 
-def pMap (E S: Type) [Stream S] (M: Type -> Type) (U V: Type) (f: U -> V) (x: ParsecT E S M U) : ParsecT E S M V  :=
+private def run_cok (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (a: A) (s₁: State S E) (_h: Hints (Stream.Token S)): M (Reply S E A) :=
```
Comment (johnchandlerburnham) [inline Megaparsec/Megaparsec.lean:81]: I don't like `private` defs, it feels too OOP-y. Many times in Haskell I've wanted to access a definition, but because it wasn't exported, I have no choice but to copy/paste code

File: Megaparsec/Megaparsec.lean [hunk 2]
```diff
@@ -58,21 +60,37 @@ inductive Result (S E A: Type) [Stream S] where
...
+  pure $ Reply.mk s₁ True (Result.ok a)
+private def run_cerr (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (err: ParseError S E) (s₁: State S E): M (Reply S E A) :=
```
Comment (johnchandlerburnham) [inline Megaparsec/Megaparsec.lean:83]:
```lean
def run_cerr {E S: Type} [Stream S] {M: Type -> Type} [Monad M] {A: Type} (err: ParseError S E) (s₁: State S E): M (Reply S E A)
```

will work to get rid of `E S M A` explicit args in `pMap`, but I don't know if it has other implications. We also might want to group type params and typeclass args, putting the Stream and Monad constraint after `A : Type`

Comment (cognivore) [reply Megaparsec/Megaparsec.lean:83]: Agree with both in principle, strongly disagree in practice in this particular case.

 2. It's cool to have constraints in one place, but jumping hither and thither is certainly not. This function head reads well in plain English "takes type E, a stream S, a monad M and a type A", rather than "takes types E S M and A, where S is a Stream and M is a Monad".

File: Megaparsec/Megaparsec.lean [hunk 3]
```diff
@@ -58,21 +60,37 @@ inductive Result (S E A: Type) [Stream S] where
...
+  pure $ Reply.mk s₁ True (Result.err err)
+private def run_eok (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (a: A) (s₁: State S E) (_h: Hints (Stream.Token S)): M (Reply S E A) :=
+  pure $ Reply.mk s₁ False (Result.ok a)
+private def run_eerr (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (err: ParseError S E) (s₁: State S E): M (Reply S E A) :=
+  pure $ Reply.mk s₁ False (Result.err err)
+def runParsecT (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (x: ParsecT E S M A) (s₀: State S E): M (Reply S E A) :=
+  -- I bet there's a way to apply types to a list of four functions with Applicative or something, but it's good enough for the time being.
+  x.unParser (Reply S E A) s₀ (run_cok E S M A) (run_cerr E S M A) (run_eok E S M A) (run_eerr E S M A)
```
[the remark this answers is not in the document]
Comment (arthurpaulino) [reply Megaparsec/Megaparsec.lean:91]: More on `private`: it's not very useful when you want to reason and then you get stuck because you can't unfold a `private` declaration.

**Comment by cognivore** [comment]
Agree with both in principle, strongly disagree in practice in this
particular case.

 1. In principle, yes, but not in this particular case. There's no `where`
clause in Lean, nor general `let in`. I could use "monadic" let, but it is
not a general way to do this. None of these four functions make much sense
in a vacuum, so I'd either `private` them or inline.

 2. It's cool to have constraints in one place, but jumping hither and
thither is certainly not. This function head reads well in plain English
"takes type E, a stream S, a monad M and a type A", rather than "takes
types E S M and A, where S is a Stream and M is a Monad".
